<h1 id="objectives">Objectives</h1>
<p>The learning objectives for this practical are:</p>
<ul>
<li>Implement a program in Python that decides if two given numbers are relatively prime.</li>
<li>Implement a program in Python that decides if a given number is cool.</li>
<li>Bundling lines together into functions.</li>
<li>How to take arguments from the command line.</li>
<li>Reading DNA from FASTA files.</li>
<li>Make syntax errors in Python.</li>
<li>Correct syntax errors in Python.</li>
<li>Debug your program when it doesn’t work.</li>
</ul>
<p>For the problems in which you have to implement a program that solves some arithmetic problem that we have solved in class at the blackboard, you should have already one working solution from that class on paper. If you don’t, please ask a colleague for one. Try to focus the time of this practical in addressing the technical challenges of running a Python program and correcting run-time errors.</p>
<p>Whenever you are stuck with an error, please consult the section entitled “Debugging” from <a href="/practical4#debugging">practical 4</a>.</p>
<h1 id="setup-and-background">Setup and background</h1>
<p>To do this practical you need an installation of Python version 3. You can find the instructions in the <a href="/setup#python-and-jupyter-notebook/">setup</a> link on how to install Python version 3 in your system. Once Python is installed, you should be able to call it from the shell in the terminal window. You can check whether that is possible by typing:</p>
<pre><code>$ which python
$ python --version</code></pre>
<p>It may happen that you have two Python installations, one corresponding to version 2.x and another to version 3.x. In that situation the previous command may say that your Python version is 2.x and to access the version 3 you need to call the executable <code>python3</code>. Try then for instance:</p>
<pre><code>$ python3 --version</code></pre>
<p>If this is your case, then whenever the executable <code>python</code> is invoked in the rest of this practical, please use <code>python3</code> instead. Make a directory for this practical, called <code>practical5</code>, and store there the files generated during the practical.</p>
<h1 id="relatively-prime-numbers">Relatively prime numbers</h1>
<p>The Wikipedia <a href="https://en.wikipedia.org/wiki/Coprime_integers">page</a> for relatively prime numbers says:</p>
<blockquote>
<p>In number theory, two integers a and b are relatively prime, mutually prime, or coprime if the only positive integer that evenly divides (is a divisor of) both of them is 1. One says also a is prime to b or a is coprime with b. Consequently, any prime number that divides one of a or b does not divide the other. This is equivalent to their greatest common divisor (gcd) being 1.</p>
</blockquote>
<p>Implement a program in Python that asks for two positive integer numbers and says whether they are relatively prime or not, providing some message with the <code>print()</code> function.</p>
<h1 id="cool-numbers">Cool numbers</h1>
<p>We say that a positive integer number <code>x</code> is <strong>cool</strong> if this number is equal to the sum of any sequence of increasing consecutive positive integer numbers smaller or equal than <code>x</code>, starting on 1. For instance, number 1 is cool because the only positive integer number smaller or equal than 1 is 1 itself; number 2 is not cool because neither 1 nor 1+2 are equal to 2; number 3 is cool because 3=1+2. Until 10, numbers 4, 5, 7, 8 and 9 are not cool. However, 6=1+2+3 and 10=1+2+3+4 are indeed cool.</p>
<p>Implement a program in Python that asks for one positive integer number and says whether the given number is cool or not, providing some message with the <code>print()</code> function.</p>
<h1 id="bundling-lines-together-into-functions">Bundling lines together into functions</h1>
<p>Programming instructions performing a specific task, such as the calculation of a particular value or decision, can be bundled together under a so-called <a href="https://en.wikipedia.org/wiki/Subroutine">function</a>; see the slides on functions from <a href="https://funcompbio.github.io/lecture5/#11">this lecture</a>.</p>
<p>Edit the previous two programs and, in each of them, bundle together the code into a function called <code>main()</code> and put a call to that function <code>main()</code> at the bottom of the file, i.e.:</p>
<pre><code>def main() :
  ## CODE OF YOUR PYTHON PROGRAM
  ## INDENTED TO BECOME BUNDLED
  ## INTO A FUNCTION CALLED main()
  ## THIS IS THE SO-CALLED &quot;BODY&quot;
  ## OF THE FUNCTION

main() ## THIS IS THE &quot;CALL&quot; TO THE FUNCTION main()</code></pre>
<p>Verify that the program runs in exactly the same way as before you did these modifications. Now edit each of the programs again with the following two new updates:</p>
<ol type="1">
<li>Move the instructions that read the input outside the body of the <code>main()</code> function and place them right before the call to the <code>main()</code> function.</li>
<li>Parametrize the <code>main()</code> function with the variables that store the input, i.e., writing <code>main(x)</code> if <code>x</code> were the variable that stores the input. At this point, verify that the program runs in exactly the same way as before.</li>
<li><p>Replace the calls to the <code>print()</code> function, printing the result message, by an assignment of that result message to a variable called <code>res</code>. Insert the following line as last line in the body of the <code>main()</code> function:</p>
<pre><code>return(res)</code></pre></li>
<li><p>Replace the call to the function <code>main()</code> by:</p>
<pre><code>print(main(x))</code></pre>
<p>assuming that <code>x</code> is the name of the variable that stores the input.</p></li>
</ol>
<p>Verify that the program runs in exactly the same way as before these modifications. Try to understand the flow of the input information into the code that makes the actual calculations and the flow of the output from the <code>main()</code> function to the screen.</p>
<h1 id="how-to-take-arguments-from-the-command-line.">How to take arguments from the command line.</h1>
<p>Programs such as Unix commands, may take the so-called <a href="https://en.wikipedia.org/wiki/Command-line_interface#Arguments">command-line arguments</a>, which allow programs to read different input each time we run them For instance, the <code>cp</code> file-copying command requires two arguments, the file to copy and the target directory or file where it should be copied, for instance:</p>
<pre><code>$ cp ~/Download/rawData.zip .</code></pre>
<p><a href="https://en.wikipedia.org/wiki/Command-line_interface#Arguments">Command-line arguments</a> allow us to re-use programs with different input and enable automatizing <a href="https://en.wikipedia.org/wiki/Workflow">workflows</a>. We will see in this section how can we take command-line arguments through the code in our own Python programs.</p>
<p>We can use the Python module <code>sys</code> (to refresh the concept of a Python module, see the corresponding section in <a href="https://funcompbio.github.io/lecture5/#22">this lecture</a>) to fetch argument values given by the user in the command line. To illustrate this functionality, create a text file called <code>showargs.py</code> and add the following lines:</p>
<pre><code>import sys
i = 0
while i &lt; len(sys.argv) :
  print(f&quot;argument vector position {i}: {sys.argv[i]}&quot;)
  i = i + 1</code></pre>
<p>If you call this program from the Unix command line with the arguments <code>give me 5</code>, you should be getting the following output:</p>
<pre><code>$ python showargs.py give me 5
argument vector position 0: showargs.py
argument vector position 1: give
argument vector position 2: me
argument vector position 3: 5</code></pre>
<p>Try to understand what are the contents of the vector <code>argv</code> in the module <code>sys</code> in the previous execution mode. Once you have understood that, try to complete the following exercise.</p>
<p>Make a new version of the Python program that decides whether two positive integer numbers are relatively prime with the following characteristics:</p>
<ol type="1">
<li><p>Instead of asking the user to enter two values, it should take them from the Unix command-line arguments.</p></li>
<li><p>The code doing the actual calculation on whether two given numbers are relatively prime should be encapsulated into a function called <code>main(x, y)</code> with two arguments <code>x</code> and <code>y</code> corresponding to the two numbers to be evaluated; see the slides on functions from <a href="https://funcompbio.github.io/lecture5/#18">this lecture</a>. The function <code>main(x, y)</code> should return a character string value set to <code>yes</code> when <code>x</code> and <code>y</code> are relatively prime and <code>no</code> when they are not. This function <strong>should not</strong> print anything on the screen.</p></li>
<li><p>When the program is called from the command line without the two arguments corresponding to the two positive integer values to evaluate, then for a program called, e.g., <code>relprime.py</code>, the user should get a message like:</p>
<pre><code>$ python relprime.py
error: relprime.py &lt;x&gt; &lt;y&gt;</code></pre></li>
<li><p>When the program is properly called from the command line with the two arguments corresponding to the two positive integer values to evaluate, then the program should call the previously defined <code>main(x, y)</code> function, take its result and print it in the terminal screen.</p></li>
</ol>
<h1 id="reading-dna-from-fasta-files">Reading DNA from FASTA files</h1>
<p>Following the explanation from Wikipedia, in bioinformatics the <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA format</a> is a text-based file format for representing sequences of either <a href="https://en.wikipedia.org/wiki/Nucleic_acid_sequence">nucleic acids</a> or <a href="https://en.wikipedia.org/wiki/Amino_acid">amino acids</a>. Here, we want to learn how can we read from a Python program the DNA sequence of a gene stored in FASTA format within a text file and make some simple summaries of the DNA content from that gene. We are going to illustrate that task with the <a href="https://en.wikipedia.org/wiki/Hemoglobin_subunit_beta"><em>hemoglobin subunit beta</em> (HBB) gene</a>, which is a gene coding for a protein that forms part of <a href="https://en.wikipedia.org/wiki/Hemoglobin">haemoglobin</a>, the molecule responsible for transporting oxygen in <a href="https://en.wikipedia.org/wiki/Red_blood_cell">red blood cells</a> of almost all <a href="https://en.wikipedia.org/wiki/Vertebrate">vertebrates</a>.</p>
<p>First, let’s download the FASTA file of the DNA of the human <em>HBB</em> gene following these steps:</p>
<ol type="1">
<li>Go to the NCBI web page for the human <em>HBB</em> gene at this <a href="https://www.ncbi.nlm.nih.gov/gene/3043">link</a>.</li>
<li><p>Scroll down to the section entitled “Genomic regions, transcripts and products” and click on the link called <code>FASTA</code>, as indicated by the red arrow in the following image.</p>
<p><img src="ncbigene1.jpg" /></p></li>
<li><p>In the next page, click on the link called <code>Send to:</code>, from the pull-down menu click on <code>File</code> and finally on the button <code>Create File</code>, as indicated by the red arrows in the following image.</p>
<p><img src="ncbigene2.jpg" /></p></li>
<li>A file called <code>sequence.fasta</code> will be downloaded from your browser, probably into your downloads directory. Move it into the folder that you are using for this practical under the name <code>HBB.fa</code>. This is the FASTA file containing the DNA of the human <em>HBB</em> gene.</li>
<li><p>Examine the first lines of this FASTA file with the Unix <code>head</code> command. The result should be as follows:</p>
<pre><code>$ head HBB.fa
&gt;NC_000011.10:c5227071-5225464 Homo sapiens chromosome 11, GRCh38.p14 Primary Assembly
ACATTTGCTTCTGACACAACTGTGTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGA
GGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGGATGAAGTTGGTGGTGAGGCCCTGGGC
AGGTTGGTATCAAGGTTACAAGACAGGTTTAAGGAGACCAATAGAAACTGGGCATGTGGAGACAGAGAAG
ACTCTTGGGTTTCTGATAGGCACTGACTCTCTCTGCCTATTGGTCTATTTTCCCACCCTTAGGCTGCTGG
TGGTCTACCCTTGGACCCAGAGGTTCTTTGAGTCCTTTGGGGATCTGTCCACTCCTGATGCTGTTATGGG
CAACCCTAAGGTGAAGGCTCATGGCAAGAAAGTGCTCGGTGCCTTTAGTGATGGCCTGGCTCACCTGGAC
AACCTCAAGGGCACCTTTGCCACACTGAGTGAGCTGCACTGTGACAAGCTGCACGTGGATCCTGAGAACT
TCAGGGTGAGTCTATGGGACGCTTGATGTTTTCTTTCCCCTTCTTTTCTATGGTTAAGTTCATGTCATAG
GAAGGGGATAAGTAACAGGGTACAGTTTAGAATGGGAAACAGACGAATGATTGCATCAGTGTGGAAGTCT</code></pre></li>
</ol>
<p>Next, create a new text file called <code>tallynt.py</code> with the following Python program:</p>
<pre><code>f = open(&quot;HBB.fa&quot;)
line = f.readline()                  ## read the first line from HBB.fa
print(&quot;The DNA sequence from gene:&quot;)
print(line.strip())                  ## print the first line from HBB.fa

seq = &quot;&quot;                             ## seq will store the whole gene DNA
while (line) :                       ## while &#39;line&#39; is not empty
    line = f.readline()              ## read the next line from HBB.fa
    seq = seq + line.strip()         ## concatenate that line to &#39;seq&#39;

v = list(seq)                        ## convert &#39;seq&#39; into a vector &#39;v&#39;
n = len(v)                           ## calculate the length of vector &#39;v&#39;
print(f&quot;has a total of {n} nucleotides&quot;)</code></pre>
<p>This Python program reads the lines from <code>HBB.fa</code> and concanates them into a single character string called <code>seq</code>, which afterwards is converted into a vector <code>v</code>, from which we calculate its length, corresponding to the total number of nucleotides in the DNA encoding the gene <em>HBB</em>. It includes a call to the following functions:</p>
<ul>
<li><a href="https://www.w3schools.com/python/ref_func_open.asp"><code>open()</code></a>, which opens a text file and returns an object that allows one to read lines from that file using the method <a href="https://www.w3schools.com/python/ref_file_readline.asp"><code>f.readline()</code></a>, for some file object <code>f</code>.</li>
<li><a href="https://www.w3schools.com/python/ref_string_strip.asp"><code>str.strip()</code></a>, where <code>str</code> is some character string object and it removes any leading and trailing spaces and newlines from that string.</li>
</ul>
<p>When you run this program, the output must be as follows:</p>
<pre><code>$ python tallynt.py
The DNA sequence from gene:
&gt;NC_000011.10:c5227071-5225464 Homo sapiens chromosome 11, GRCh38.p14 Primary Assembly
has a total of 1608 nucleotides</code></pre>
<p>Notice that you could also count the nucleotides using a Unix shell command line as follows:</p>
<pre><code>$ grep -v &#39;&gt;&#39; HBB.fa | fold -1 | wc -l
1609</code></pre>
<p>Try to understand what is doing each bit of the previous Unix command-line. Could you figure out why this command line is giving a number that is one unit larger than what our Python programe gives?</p>
<p>Next, improve the Python program with the following two enhancements:</p>
<ol type="1">
<li><p>Enable the program to take the name of the FASTA file as a first argument in the command line, so that it can work with any FASTA file containing DNA from any gene. Run it, for instance, with the DNA from the the <a href="https://www.ncbi.nlm.nih.gov/gene/59272">angiotensin converting enzyme 2 (<em>ACE2</em>) gene</a>, which encodes a protein that acts as a <a href="https://www.nytimes.com/interactive/2020/03/11/science/how-coronavirus-hijacks-your-cells.html">receptor for the spike glycoprotein of the human coronavirus SARS-CoV-2</a>, the causative agent of coronavirus disease-2019 (COVID-19), e.g.:</p>
<pre><code>$ python tallynt.py ACE2.fa</code></pre></li>
<li><p>Enable the program to take a second argument where we specify one of the four possible nucleotides (<code>A</code>, <code>C</code>, <code>G</code> or <code>T</code>) and it calculates the number of occurrences of that nucleotide in the DNA sequence of the gene, e.g.:</p>
<pre><code>$ python tallynt.py HBB.fa A</code></pre>
<p>You can verify whether the calculation is correct by doing it also in Unix with the following command line:</p>
<pre><code>$ grep -v &#39;&gt;&#39; HBB.fa | fold -1 | grep A | wc -l
411</code></pre>
<p>Try to understand what is doing each bit of the previous Unix shell command line. Could you think of a way to tally all four nucleotides in the same command line?</p></li>
</ol>
